### PR TITLE
Add new mushroom base terrain and make the overlay have mixed defense

### PR DESCRIPTION
**Pull request by CelticMinstrel**
Addresses #4213 / #4214 since no-one else seemed interested in doing it. Graphics courtesy of @doofus-01.

Probably needs work on the transitions. I started with the original (now-reverted) commit and tweaked it to create new terrains instead of modifying the old terrain. Also, if anyone has a better terrain code than `Tb`, please suggest it.

I didn't fiddle with any existing maps, as I'm not sure whether it would be preferred to replace `*^Uf` with `Tb` or with `*^Tf`. That should be left to someone who cares about MP balance.

### CODE REVIEW

**Comment by nemaara** [comment]
Do you have a screenshot of the before and after? What is this change supposed to look like?

File: data/core/terrain.cfg [hunk 1]
```diff
@@ -2721,7 +2758,7 @@ For those who go by land or sea, a bridge is the best of both worlds — for gam
     id=fungus
     name= _ "Fungus"
     editor_name= _ "Fungus"
-    string=Uft
+    string=Tt
```
Comment (jostephd) [inline data/core/terrain.cfg]: This seems to undefine the `Uft` terrain code. Is that so? Is there a backward compatibility problem here?

**Comment by CelticMinstrel** [comment]
It's not supposed to change appearance of any existing terrains. I'll grab a screenshot at some point, but probably tomorrow as it's late and I should sleep.

File: data/core/terrain.cfg (hunk 1 above)
Comment (CelticMinstrel) [reply data/core/terrain.cfg]: The `Uft` code isn't a code used on maps, it's the "archetype" used only as a base for other terrains and to provide general information about that archetype (like the sidebar icon and help text). As you can see, it's marked `hidden`, and furthermore it doesn't actually have any graphics, so unless someone manually edited it into their map, there shouldn't be any issues.

**Comment by jostephd** [comment]
Thanks for working on this. I do have a screenshot, but it doesn't look quite right:

2019-08-31-035237_278x267_scrot

Pressing Ctrl+T on the first row opens the help at the homepage.

Pressing Ctrl+T on the second row opens the terrain help page, the terrain image shows mushrooms on grass. It's a mixed terrain.

Pressing Ctrl+T on the third row opens a more generic terrain description page, there's no image nor a list of stats, it's not a mixed terrain.

**Comment by nemaara** [comment]
I don't personally use mixed mushroom terrains for their unit movement/defense properties, so balancing around the new behavior isn't going to be an issue for me at least. Since it doesn't change the appearance of existing terrains, I'm totally fine with this.

**Comment by CelticMinstrel** [comment]
Well, that's why I marked it WIP. I'll take a look at the issue tomorrow.

**Comment by jostephd** [comment]
> Since it doesn't change the appearance of existing terrains, I'm totally fine with this.

If it doesn't change the appearance of existence terrains, it shouldn't claim to address #4214.

**Comment by CelticMinstrel** [comment]
It addresses #4214 by superceding it in my view? But if you disagree, then whatever.

Anyway, looks like it works now. A lot of the problems were caused by conflicting terrain IDs (oops!).

**Comment by jostephd** [comment]
> It addresses #4214 by superceding it in my view? But if you disagree, then whatever.

I'm not trying to nitpick. #4214 is about changing the graphics of `^Uf` on master to match its stats. This PR doesn't change either the graphics or the stats of `^Uf`, so it doesn't fix #4214. I was just trying to clarify this point. Let's move on.

File: data/core/terrain.cfg (hunk 1 above)
Comment (stevecotton) [reply data/core/terrain.cfg]: If someone's cut&pasted terrain definitions in to their UMC, then those terrains might be using Uft as a base. In 1.14, it seems that all terrain aliases use only the virtual terrains as their bases, and I think the help pages start doing odd things if you don't (as in, the game creates a help section for every terrain used as a base, even if it's not a virtual terrain).

**Comment by CelticMinstrel** [comment]
One minor detail I was wondering is whether the default base of the new mushrooms should be `Uu` (as with the old mushrooms) or `Tb`.

**Comment by Vultraz** [comment]
Honestly, I do not find this worth it at all. Like I said on Discord (just copying it), in order to solve one terrain not functioning quite as expected, we're proposing deprecating the whole terrain and introducing a new one with a full set of graphics, necessitating all mainline and UMC maps that use the mushrooms (including in non-cave areas!) be updated. The new terrain proposed, with it's cave floor base, doesn't work in outdoor maps.

This is not just a graphics change, like when we went from smooth cave walls to mine walls. That change did result in some aesthetic differences but the general purpose was the same. This changes the use of the terrain completely.

And what do we get out of it? Sure, the terrain becomes less confusing and more consistent. But the nominal gain for the amount of work just does not seem worth it. At all.

**Comment by CelticMinstrel** [comment]
>The new terrain proposed, with it's cave floor base, doesn't work in outdoor maps.

Uh. What the heck are you talking about? This statement seems to be in disagreement with reality...

**Comment by CelticMinstrel** [comment]
@Vultraz
>The new terrain proposed, with it's cave floor base, doesn't work in outdoor maps.

Uh. What the heck are you talking about? This statement seems to be in disagreement with reality...

1. It doesn't have a cave floor base, it has a mushroom floor base.
2. You can put it on whatever base you want.
3. Exactly why wouldn't it work in outdoor maps?

**Comment by Vultraz** [comment]
I meant now you can't put it on top of hills, mountains, swamp, etc.

**Comment by CelticMinstrel** [comment]
Uh... yes you can?

**Comment by stevecotton** [comment]
> One minor detail I was wondering is whether the default base of the new mushrooms should be `Uu` (as with the old mushrooms) or `Tb`.

I think Tb - that way, when updating maps that have ^Uf terrains on them:
* left-click places terrain which is functionally the same
* shift-left-click places terrain which is cosmetically the same

I'm currently playing through Eastern Invasion, updating the maps.

**Comment by Vultraz** [comment]
Is it still the same giant mushrooms with a new optional base?

**Comment by CelticMinstrel** [comment]
There are three new terrains in this PR:

1. `Tb`, which is a base terrain with mushroom properties
2. `^Tf`, which is visually equivalent to `^Uf` and aliases to mushrooms, but has mixed properties like forest.
3. `^Tfi`, which is the illuminated version of above, like `^Ufi`.

**Comment by Pentarctagon** [comment]
Sounds good to me as well, FWIW.

File: data/core/terrain.cfg (hunk 1 above)
Comment (CelticMinstrel) [reply data/core/terrain.cfg]: And wmllint could add rules to auto-update terrain definitions that alias to `Uft`,

**Comment by CelticMinstrel** [comment]
Oh, I just noticed @jostephd's observation about Ctrl+T... I never thought to test that myself. Did you ever check up on it again after I fixed all the other issues?

**Comment by jostephd** [comment]
> Oh, I just (re-)noticed @jostephd's observation about Ctrl+T... I never thought to test that myself. Did you ever check up on it again after I fixed all the other issues?

I did not test it again.

**Comment by CelticMinstrel** [comment]
Okay, I just checked... Ctrl+T works for the overlay and shows the correct topic depending on the base terrain, but for some reason it doesn't work for the base terrain, even if I put a different overlay on it.

I don't think it's a blocker though.

**Pull request merged**